### ISSUE TITLE
Service Layer?

### ISSUE THREAD

**Issue by GrahamCampbell** [cachethq/Cachet#189]
I service layer would be good to introduce now while the codebase is still small. I'm not saying we change to all out ddd, but a small service layer would definitely be a good idea.

**Comment by jbrooksuk** [comment]
Yeah, I think this would be worth while. Are you ok doing this for us, or just leaving it as a suggestion?

**Comment by GrahamCampbell** [comment]
Leaving it as a suggestion for now. Sorry, I've got no time at the moment. :(

**Comment by jbrooksuk** [comment]
Ok, no problem. I need to learn more about the Service Layer pattern before I implement it.

**Comment by GrahamCampbell** [comment]
Ross tuck did a really good talk about it at drupalcon this year. It's on youtube.

**Comment by GrahamCampbell** [comment]
He does go really deep in the second half, and that stuff isn't really appropriate here because I don't see it working if we want to continue using eloquent.

**Comment by GrahamCampbell** [comment]
So actually, watching the first 20 mins should give you a good idea of what you want to do. :)

**Comment by jbrooksuk** [comment]
Awesome, I'll take a look when I get a bit of time :)

**Comment by jbrooksuk** [comment]
Oh and [this](https://www.youtube.com/watch?v=3uV3ngl1Z8g) is the video @GrahamCampbell is talking about.

**Comment by GrahamCampbell** [comment]
No.

**Comment by jbrooksuk** [comment]
Oh?

**Comment by GrahamCampbell** [comment]
https://www.youtube.com/watch?v=ajhqScWECMo

**Comment by jbrooksuk** [comment]
Ah sorry, the drupalcon one.

**Comment by GrahamCampbell** [comment]
yeh :)

**Comment by joecohens** [comment]
:+1:

**Comment by GrahamCampbell** [comment]
It's very important that we keep the http logic inside the controllers. I'm already seeing calls to the request and input facades from else where - even in the views. This shouldn't be happening.

**Comment by GrahamCampbell** [comment]
We can't even start the service layer if we haven't got that stuff clear first.

**Comment by joecohens** [comment]
I know we first have to move all the views logic to the controller and then start a service layer.

**Comment by GrahamCampbell** [comment]
Yeh, I'll be doing that before I move the models into the src folder and namespace them.

**Comment by GrahamCampbell** [comment]
The stuff on the right is what we want. :)

<img src="https://cloud.githubusercontent.com/assets/2829600/5593255/4e715528-91fa-11e4-884a-1cc5dbd557cf.png" align="left" height="400" >

**Comment by joecohens** [comment]
:+1: I use [ComponentService, ComponentCreatorService, ComponentUpdaterService] naming on some projects for the services what do you think?

**Comment by GrahamCampbell** [comment]
Hmm, services aren't meant to be as clean cut as that. It's easy to fall into that trap, the same with repositories, and they we just end up with piles of nested classes, which isn't really helping either.

**Comment by joecohens** [comment]
Ok ok but it's also easy to have a ComponentService bloated because the name it's too general.

**Comment by GrahamCampbell** [comment]
Yeh, but it would still be less bloated than our controllers would be otherwise, right?

**Comment by GrahamCampbell** [comment]
I find services are more useful for abstracting other functionality, rather than that core read/write stuff.

**Comment by joecohens** [comment]
That's right! Let's go one step at the time :)

**Comment by GrahamCampbell** [comment]
Another step towards this then: #222. :)

**Comment by GrahamCampbell** [comment]
Ok, the next refactor is to namespace the models. I'm actually going to alias the Setting model like a facade so we can still use it in our views without any changes for now.

I'll start this in about an hour's time. :)

**Comment by joecohens** [comment]
Excellent!

**Comment by jbrooksuk** [comment]
Ooh, aliasing the `Setting` model is something that sounds like something I need to see!

**Comment by GrahamCampbell** [comment]
Yeh, it'll be a temporary hack for now.

**Comment by GrahamCampbell** [comment]
Also, the `Service` model seems unused at the moment?

**Comment by GrahamCampbell** [comment]
And so does the `Subscriber` model.

**Comment by jbrooksuk** [comment]
Yeah:

- `Service` will be used to fetch data about a third-party integration.
- `Subscriber` is for fetching a list of people who have subscribed to reports. They'll receive an email for every component and incident change.

**Comment by GrahamCampbell** [comment]
And the webhooks stuff is totally unused too. Also, I don't think that stuff is really suitable for a model.

**Comment by jbrooksuk** [comment]
@GrahamCampbell web hooks are not fully implemented yet, but they'll essentially `POST` data to other services for them to act upon. How would you suggest doing a web hook? There will still need to be two models for handling the database though? Maybe making this into a package?

**Comment by GrahamCampbell** [comment]
A service class would be were all the guzzle stuff happens.

**Comment by GrahamCampbell** [comment]
Also, we can discuss that later. I'd like to focus on refactoring the stuff we currently have for now.

**Comment by joecohens** [comment]
:+1: One step at the time haha.

**Comment by jbrooksuk** [comment]
Of course, I was just interested in how we'd do it. It makes sense to leave that until the "under the hood stuff" is done, then we can move forward with new features.

I'm holding off on #196 and #197 until your refactor is complete :)

**Comment by joecohens** [comment]
I'm interested too :)

**Comment by GrahamCampbell** [comment]
Yeh, this refactor is a big one...

**Comment by jbrooksuk** [comment]
Take your time, it'll be worth it!

**Comment by jbrooksuk** [comment]
I'm going to be studying the code to really understand all of this cool stuff!

**Comment by joecohens** [comment]
:+1:

**Comment by GrahamCampbell** [comment]
Yeh, it's taking so long because I'm doing two other things at the same time, lol...

**Comment by jbrooksuk** [comment]
It's only been half an hour ;)

**Comment by joecohens** [comment]
Lol!

**Comment by GrahamCampbell** [comment]
#235. Not quite done yet.

**Comment by joecohens** [comment]
@GrahamCampbell Awesome!! This will also helps us move to Laravel 5 faster :)

**Comment by GrahamCampbell** [comment]
Yeh, definately.

**Comment by GrahamCampbell** [comment]
I think this redactor might be a good idea once we've actually moved to laravel 5. NB, I'd be happy to get us migrated to laravel 5 in early February, then get started on this stuff.

**Comment by joecohens** [comment]
I think so too! And that would be great ;)

**Comment by jbrooksuk** [comment]
Oooh, we're planning Laravel 5, I like it!

**Comment by GrahamCampbell** [comment]
Laravel 5 has some really cool stuff in it. Once it's released in stable, and dingo is ready, I'll upgrade us  to it. I'll create a specific issue regarding the new plan.

**Comment by jbrooksuk** [comment]
You. Are. Awesome.

Thanks Graham!

**Comment by GrahamCampbell** [comment]
Pleasure. :)
